Fixed. #4の課題が完了しました。

TsunakiM opened the pull request.

TsunakiM committed:
[WIP]#4の課題に着手しました。

TsunakiM committed:
途中経過。１人だけなら表示できるようになった。

TsunakiM committed:
入力をまとめた

TsunakiM committed:
配列への追加が上手く行っていないケース

TsunakiM committed:
配列への追加が上手く行っていないケース２

TsunakiM committed:
配列の追加ができて、それらしき挙動になってきた。

TsunakiM committed:
Fixed. #4の課題が完了しました。

taguchi-k commented on traning1-1-4/traning1-1-4/Account.m:
```diff
@@ -0,0 +1,50 @@
+
+#import <Foundation/Foundation.h>
+#import "Account.h"
+
```
可読性のために一行空白行をつけるのはOKですが、2行つける必要はないので1行削除してください。
他の空白行が二行になっているところも同様です。

taguchi-k commented on traning1-1-4/traning1-1-4/Account.h:
```diff
@@ -0,0 +1,25 @@
+// プロパティの設定
+@interface Account : NSObject
+
+@property (nonatomic)NSString *name;
```
`name`〜`favLang `まで、外部から利用しないのであれば実行ファイル側で定義してください。
外部からの思わぬ値の変更を防ぐためです。
外部から利用する必要があるのであればその限りではありません。

taguchi-k commented on traning1-1-4/traning1-1-4/Account.h:
```diff
@@ -0,0 +1,25 @@
+@property (nonatomic)NSUInteger age;
+@property (nonatomic)NSString *sex;
+@property (nonatomic)NSString *favLang;
+@property (nonatomic)NSMutableArray *employeeList;
```
`Account`クラスに社員一覧のプロパティは不要です。
実装ファイル側で詳細コメントします。

taguchi-k commented on traning1-1-4/traning1-1-4/Account.m:
```diff
@@ -0,0 +1,50 @@
+
+@implementation Account
+
+@synthesize name, age, sex, favLang, employeeList;
```
コンパイラが補完してくれるため`@synthesize`は不要です。

taguchi-k commented on traning1-1-4/traning1-1-4/Account.m:
```diff
@@ -0,0 +1,50 @@
+
+@synthesize name, age, sex, favLang, employeeList;
+
++(Account*)Account {
```
インスタンスを初期化したいだけであれば、メソッドを新たに作る必要はなく`alloc`して`init`すればOKです。

```
Account *account = [[Account alloc] init];
```

taguchi-k commented on traning1-1-4/traning1-1-4/Account.m:
```diff
@@ -0,0 +1,50 @@
+    return obj;
+}
+
+-(void)addEmployee :(NSString*)NAME :(NSUInteger)AGE :(NSString*)SEX :(NSString*)FAVLANG{
```
引数名はLowerCamelCaseにしてください。
https://github.com/SmartTechVentures/Objective-C-style-guide

taguchi-k replied:
今回の場合であればプロパティへの値のセットはイニシャライザ作ってそこでセットしてあげればいいと思います。

taguchi-k commented on traning1-1-4/traning1-1-4/Account.m:
```diff
@@ -0,0 +1,50 @@
+}
+
+-(void)addEmployee :(NSString*)NAME :(NSUInteger)AGE :(NSString*)SEX :(NSString*)FAVLANG{
+    name = NAME;
```
プロパティにアクセスする際は`self`+`.`でアクセスするようにしてください。
独自のsetterやgetterが定義されている場合そちらを呼ぶようにするためです。

※プロジェクトによっては`_プロパティ名`でアクセスするルールになっているところもありますので、プロジェクトのルールにしたがってください。

taguchi-k commented on traning1-1-4/traning1-1-4/Account.m:
```diff
@@ -0,0 +1,50 @@
+    sex = SEX;
+    favLang = FAVLANG;
+
+    [employeeList addObject:@[NAME, @(AGE), SEX, FAVLANG]];
```
この処理だと`Account`型インスタンスの配列を作れていません。
各引数の要素をもった配列の配列（二次元配列）になっています。

taguchi-k commented on traning1-1-4/traning1-1-4/main.m:
```diff
@@ -0,0 +1,30 @@
+        [obj addEmployee:@"山田太郎" :24 :@"M" :@"Java"];
+        [obj addEmployee:@"佐藤大輔" :31 :@"M" :@"C#"];
+        [obj addEmployee:@"鈴木花子" :22 :@"F" :@"Javascript"];
+        [obj addEmployee:@"Jon Coner" :24 :@"M" :@"C++"];
```
これだと`obj`の各プロパティはここで設定したもののみになります。
次の処理にブレークポイントをセットするなりログを出力するなりして確認してみてください。

taguchi-k reviewed (changes requested):
@TsunakiMasuda 
レビューしましたので、確認・対応してください。

TsunakiM committed:
修正途中経過。コードを修正し、一人ずつなら表示できる状態。

TsunakiMasuda commented:
@taguchi-k 
おつかれさまです。ご指摘頂いた部分を、分かる範囲で修正してみました。
一人ずつなら表示できるのですが、この先、どうすればいいのかがどうしても分からないため、
何か方向性のアドバイスを頂ければ幸いです。

taguchi-k commented on traning1-1-4/traning1-1-4/Account.m:
```diff
@@ -0,0 +1,36 @@
+
+@implementation Account
+
+- (void)init :(NSString *)name :(NSUInteger)age :(NSString *)sex :(NSString *)favoriteLanguage {
```
`Account `型のインスタンスを返していないので、イニシャライザになっていません。
本メソッドはインスタンス化後に自身のプロパティを設定するメソッドになっています。
それでもやりたいことができないわけではないので構いませんが、それであれば`init`という命名が適切ではありません。

taguchi-k replied:
ちなみにイニシャライザはコーディング規約のところにもコードの例があるので参考にしてみてください。

taguchi-k commented on traning1-1-4/traning1-1-4/Account.m:
```diff
@@ -0,0 +1,36 @@
+    self.favoriteLanguage = favoriteLanguage;
+}
+
+- (void)printDetails {
```
本メソッドはこんな感じでいいと思います！

taguchi-k commented on traning1-1-4/traning1-1-4/main.m:
```diff
@@ -0,0 +1,26 @@
+
+    int main (int argc, const char * argv[]) {
+        
+        Account *account = [Account new];
```
ここでは`account `しか`Account `型のインスタンスがないですよね。
`Account `型のインスタンスを一覧分作って、それをリストとしてみてはいかがですか？

taguchi-k reviewed (changes requested):
@TsunakiMasuda 
ヒントも含めレビューしたので確認してください。

TsunakiM committed:
Fix. #4の修正。

TsunakiMasuda commented:
@taguchi-k
おつかれさまです。ヒントのおかげでそれっぽい形になりました。
ありがとうございます。
再度レビューをお願いいたします。

taguchi-k commented on traning1-1-4/traning1-1-4/Account.m:
```diff
@@ -0,0 +1,38 @@
+
+@implementation Account
+
+- (id)init :(NSString *)name :(NSUInteger)age :(NSString *)sex :(NSString *)favoriteLanguage {
```
【want】
引数にラベルをつけるようにするのがObjective-Cでは一般的です。
本メソッド利用時に引数に何を入れればいいのかわかりやすくなりますのでつけてあげたほうがわかりやすいです。

taguchi-k commented on traning1-1-4/traning1-1-4/Account.m:
```diff
@@ -0,0 +1,38 @@
+@implementation Account
+
+- (id)init :(NSString *)name :(NSUInteger)age :(NSString *)sex :(NSString *)favoriteLanguage {
+    self.name = name;
```
コーディング規約を参考に`self`がnilかどうかの判定を入れるようにしましょう。

taguchi-k commented on traning1-1-4/traning1-1-4/Account.m:
```diff
@@ -0,0 +1,38 @@
+    return self;
+}
+
+- (void)printDetails {
```
【want】
男性と女性のパターンを決めて、想定外の値が入っていた場合の`else`のパターンもあってもいいかなと思います。

taguchi-k commented on traning1-1-4/traning1-1-4/main.m:
```diff
@@ -0,0 +1,39 @@
+
+    int main (int argc, const char * argv[]) {
+        
+        NSMutableArray *allEmployees = [NSMutableArray array];
```
【want】
これでもいいのですが、`NSArray`で実装したほうがよりいいかなという印象です。

taguchi-k reviewed (commented):
レビューしましたので確認してください。

taguchi-k commented on traning1-1-4/traning1-1-4/main.m:
```diff
@@ -0,0 +1,39 @@
+        [employee0003 printDetails];
+
+        NSLog(@"~~~社員一覧~~~");
+        for(int i = 0; i < allEmployees.count; i++) {
```
ここは`for-in`で書けます。
パフォーマンスの観点とコードがシンプルになる観点から、可能であればこちらを使いましょう。

TsunakiM committed:
Fix. #4の課題の修正（指定イニシャライザの修正、メソッド等の適正化）

・指定イニシャライザの引数にラベル追加、selfがnilかの判定追加。
・printDetailsメソッドに例外処理を追加。
・NSMutableArrayをNSArrayに修正。それに伴い、配列作成のコードの変更。
・社員一覧をログ出力するfor文を高速列挙構文に変更。それに伴い、新規メソッドを１つ追加。

TsunakiMasuda commented:
@taguchi-k 
レビューありがとうございます。
ご指摘頂いた点を修正しましたので、ご確認をお願いいたします。

taguchi-k commented:
@TsunakiMasuda 
レビューOKです！お疲れ様でした！
マージしてissuesをクローズしてください。
